### DISCUSSION TITLE
TLS fingerprint

### DISCUSSION THREAD

**Discussion by utya** [arendst/Tasmota#15967] [Q&A] [answered]
Hello. Im using traefik (reverse proxy with LetsEncrypt) +mosquitto. So i try to connect (esp32) using tls. Command SetOption132 0 doesn't work every time error 62, but SetOption132 1 works good. But every time you should make fingerpint:

1. MqttFingerprint 0
2. recoonect to server
if device turn off/on.
  
How i can fix this problem.

**Comment by s-hadinger** [comment]
If the certificate comes from Letsencrypt it should work without fingerprint. Otherwise there is auto learning the first time. 

Can you share logs with `weblog 4` level?

**Reply by utya** [reply]
RESULT = {"WebLog":4}
16:56:31.071 WIF: Checking connection...
16:56:44.317 MQT: Attempting connection...
16:56:44.720 MQT: TLS connection error: 62
16:56:44.725 MQT: Connect failed to mqtt.fgfg.net:8883, rc -2. Retry in 30 sec
16:56:51.073 WIF: Checking connection...

so now i checking my letsencrypt cert

**Comment by s-hadinger** [comment]
I can't get the certificate from mqtt.fgfg.net:8883
Is it the actual URL?

Can you also paste the logs when using Fingerprints?

**Reply by utya** [reply]
mqtt.fir-ams.ru:8883 real

**Reply by utya** [reply]
openssl s_client -connect mqtt.fir-ams.ru:8883 -showcerts </dev/null | openssl x509 -noout -dates
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
verify return:1
depth=0 CN = mqtt.fir-ams.ru
verify return:1
notBefore=Jul 11 11:16:17 2022 GMT
notAfter=Oct  9 11:16:16 2022 GMT
DONE

**Reply by utya** [reply]
with fingerprint
20:29:54.218 -> 17:29:54.004 BLE: MasterEnable->1
20:29:54.218 -> 17:29:54.013 BLE: BLETask:Starting NimBLE
20:29:54.882 -> 17:29:55.153 MQT: Attempting connection...
20:29:55.437 -> 17:29:56.140 MQT: TLS connection error: 1
20:29:55.437 -> 17:29:56.146 MQT: Connect failed to mqtt.fir-ams.ru:8883, rc -2. Retry in 10 sec
20:29:57.061 -> 17:29:57.767 QPC: Reset
20:29:58.067 -> 17:29:58.762 RSL: SENSOR = {"Time":"2022-07-12T17:29:58","MJ_HT_V1344a42":{"mac":"582d34344a42","Temperature":30.4,"Humidity":54.7,"DewPoint":20.2,"RSSI":-85},"TempUnit":"C"}

20:31:12.582 -> 17:31:13.260 CMD: MqttFingerprint 0
.......


20:31:19.789 -> 17:31:19.004 BLE: MasterEnable->1
20:31:19.823 -> 17:31:19.013 BLE: BLETask:Starting NimBLE
20:31:20.419 -> 17:31:19.609 MQT: Attempting connection...
20:31:21.873 -> 17:31:21.133 MQT: TLS connected in 1265 ms, stack low mark 2020
20:31:21.910 -> 17:31:21.135 MQT: MFLN not supported by TLS server
20:31:21.910 -> 17:31:21.138 MQT: Fingerprint learned: FF 94 98 40 A9 DB B7 4A 96 15 54 C4 FF 5E F3 D2 A9 80 79 FE
20:31:22.783 -> 17:31:22.015 MQT: Connected
20:31:22.816 -> 17:31:22.028 MQT: torsion/tele/tasmota_0DBB44/LWT = Online (retained)

**Comment by utya** [comment]
Now I only had to do Fingerprint 0 once, and everything seems to be working fine for 5 minutes

**Comment by s-hadinger** [comment, marked as answer]
If it fails again check if the learned value for fingerprint is the same

**Reply by utya** [reply]
If I change the address, I need to call fingerprint it again, can I automate it?

**Reply by s-hadinger** [reply]
You can use fingerprint with all FFs (20 times FF) and it will accept any fingerprint which means accepting any certificate

**Reply by utya** [reply]
I think this is a good solution in the current situation, thanks

**Comment by utya** [comment]
so you see letsencrypt valid?

**Reply by s-hadinger** [reply]
Yes it is valid. It's working right outside of the box for me.

```
09:27:59.309 MQT: Attempting connection...
09:28:00.616 MQT: TLS connection error: 0
09:28:00.617 MQT: Connect failed to mqtt.fir-ams.ru:8883, rc 5. Retry in 10 sec
```

**Reply by utya** [reply]
but connection error

**Reply by s-hadinger** [reply]
TLS connection error:0 means that TLS worked. 

The latter error is because I don't have valid Mqtt credentials

**Comment by barbudor** [comment]
If going through the Internet you may need to increase MqttWifiTimeout (rc error -2)

**Comment by utya** [comment]
It works somehow not stable, it does not connect at all with SetOption132 0. with SetOption132 1 Once a day, you need to do mqttfingerprint 1
